common.ts

```diff
@@ -1,84 +1,91 @@
-export class Cursor {
-  view;
-  i = 0;
+export interface Codec<T> {
+  encode: (value: T) => Uint8Array;
+  decode: (buffer: Uint8Array) => T;
+  _staticSize: number;
+  _encode: (buffer: EncodeBuffer, value: T) => void;
+  _decode: (buffer: DecodeBuffer) => T;
+}
 
-  constructor(readonly u8a: Uint8Array) {
-    this.view = new DataView(u8a.buffer, u8a.byteOffset, u8a.byteLength);
-  }
+export function createCodec<T>(codec: Pick<Codec<T>, "_staticSize" | "_encode" | "_decode">): Codec<T> {
+  const { _staticSize, _encode, _decode } = codec;
+  return {
+    _staticSize,
+    _encode,
+    _decode,
+    encode(value) {
+      const buf = new EncodeBuffer(_staticSize);
+      _encode(buf, value);
+      return buf.finish();
+    },
+    decode(array) {
+      const buf = new DecodeBuffer(array);
+      return _decode(buf);
+    },
+  };
 }
 
-export type Native<C extends Codec> = C extends Codec<infer N> ? N : never;
+export type Native<T extends Codec<any>> = T extends Codec<infer U> ? U : never;
 
-/** A means of encoding and decoding a type `T` to and from its SCALE byte representation */
-export abstract class Codec<T = any> {
-  /** The minimum size of values encoded with this codec */
-  abstract _minSize: number;
-  /**
-   * Accepts the decoded value and returns the size minus minSize.
-   * May be overridden by subclasses.
-   */
-  _dynSize(_value: T): number {
-    return 0;
-  }
-  /**
-   * If true, _dynSize will always return 0.
-   * This may be overridden by subclasses, but it must still uphold the above contract.
-   */
-  _dynSizeZero = this._dynSize === Codec.prototype._dynSize;
-
-  /** Accepts the cursor and value and writes its bytes into the cursor's byte array */
-  abstract _encode(cursor: Cursor, value: T): void;
+export class EncodeBuffer {
+  finishedArrays: Uint8Array[] = [];
+  finishedSize = 0;
+  array: Uint8Array;
+  queuedArrays: Uint8Array[] = [];
+  view;
+  index = 0;
 
-  /** Decodes the currently-focused bytes into `T` and moves the cursor as is appropriate */
-  abstract _decode(cursor: Cursor): T;
+  constructor(size: number) {
+    this.array = new Uint8Array(size);
+    this.view = new DataView(this.array.buffer);
+  }
 
-  constructor() {}
+  add(buffer: Uint8Array) {
```

Comment:
I'm not big on these names, but I don't know of better ones. I named them `push`/`pop` because the structure is somewhat analogous to a stack of Uint8Arrays, and `add` because I didn't have a better name.

Reply:
Given that the encode/decode buffers are referenced as "buffer", do you think the method names pose some confusion for onlookers? Will they quickly determine that it's not a standard buffer? Should we change the param/reference names to something along the lines of "dBuffer" / "eBuffer"?... or leave as is?

Reply:
Maybe `push`/`pop`/`add` -> `pushAlloc`/`popAlloc`/`insertArray`

Reply:
I'm also open to a better word than "buffer"

Reply:
Buffer is fitting imo. Up to you.